lib/elixir/lib/module/types.ex

```diff
@@ -0,0 +1,850 @@
+  # */2 +/1 +/2 -/1 -/2 //2
+  # abs/2 ceil/1 floor/1 round/1 trunc/1
+  # elem/2 hd/1 in/2 length/1 map_size/1 tl/1 tuple_size/1
+  # is_function/1 is_function/2 is_list/1 is_number/1 is_pid/1 is_port/1 is_reference/1 is_tuple/1
+  # node/1 self/1
+  # binary_part/3 bit_size/1 byte_size/1 div/2 rem/2 node/0
+
+  @type_guards %{
+    is_atom: :atom,
+    is_binary: :binary,
+    is_bitstring: :binary,
+    is_boolean: :boolean,
+    is_float: :float,
+    is_integer: :integer,
+    is_map: {:map, []}
+  }
+
+  defp type_guard({{:., _, [:erlang, guard]}, _, [arg]}) do
+    case :maps.find(guard, @type_guards) do
+      {:ok, type} -> {:ok, arg, type}
+      :error -> {:error, :unknown_guard}
+    end
+  end
+
+  defp type_guard(_other), do: {:error, :unknown_guard}
+
+  # binary-pattern :: specifier
+  defp of_binary({:"::", _meta, [expr, specifiers]} = full_expr, context) do
+    specifiers = List.flatten(collect_specifiers(specifiers))
+
+    expected_type =
+      cond do
+        :integer in specifiers -> :integer
+        :float in specifiers -> :float
+        # TODO: Add bitstring type
+        :bits in specifiers -> :binary
+        :bitstring in specifiers -> :binary
+        :bytes in specifiers -> :binary
+        :binary in specifiers -> :binary
+        :utf8 in specifiers -> :integer
+        :utf16 in specifiers -> :integer
+        :utf32 in specifiers -> :integer
+        true -> :integer
+      end
+
+    expr_stack(full_expr, context, fn context ->
+      with {:ok, type, context} <- of_pattern(expr, context),
+           {:ok, _type, context} <- start_unify(type, expected_type, context),
+           do: {:ok, context}
+    end)
+  end
+
+  # binary-pattern
+  defp of_binary(expr, context) do
+    case of_pattern(expr, context) do
+      {:ok, type, context} when type in [:integer, :float, :binary] ->
+        {:ok, context}
+
+      {:ok, type, _context} ->
+        {:error, {:invalid_binary_type, type}}
+
+      {:error, reason} ->
+        {:error, reason}
+    end
+  end
+
+  # Push expr to stack inside fun and pop after fun
+  # The expression stack is used for
+  defp expr_stack(expr, context, fun) do
+    expr_stack = context.expr_stack
+
+    case fun.(%{context | expr_stack: [expr | context.expr_stack]}) do
+      {:ok, context} -> {:ok, %{context | expr_stack: expr_stack}}
+      {:ok, type, context} -> {:ok, type, %{context | expr_stack: expr_stack}}
+      {:error, reason} -> {:error, reason}
+    end
+  end
+
+  defp start_unify(left, right, context) do
+    case unify(left, right, context) do
+      {:ok, type, context} -> {:ok, type, %{context | unify_stack: []}}
+      {:error, reason} -> {:error, reason}
+    end
+  end
+
+  defp unify(type, {:var, var}, context) do
+    case :maps.get(var, context.types) do
+      :unbound ->
+        context = refine_var(var, type, context)
+        context = push_unify_stack(var, context)
+
+        if recursive_type?(type, [], context) do
```

Comment:
Can we do this check less frequently? Since `refine_var` stores the traces per var, you can only have a recursion if the trace for the var was previously set and you are visiting it again.

Reply:
That's a good idea, but I am not sure about using the tracing data for the algorithm. Do you think that's okay?